### PR TITLE
[ONNX] Handle Sub-block index_put in _jit_pass_onnx_remove_inplace_ops_for_onnx

### PR DESCRIPTION
**Pull request by jiafatom**
For the added UT and existing UTs, this code is independent and ready for review.

### CODE REVIEW

**Review by BowenBao — COMMENTED** [review]
some inline comments. It would be helpful if each step there is some comments with simple IR snippets showing before and after comparison

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 1]
```diff
@@ -336,7 +336,64 @@ void SquashSliceAndSelect(Node* index_put_node) {
...
+    next_node->addOutput()->copyMetadata(new_index_put);
+    next_block = next_node->owningBlock();
+    node_list.emplace_back(std::make_pair(outer_block, next_node));
+  }
+
+  // Register index_put inputs through the blocks.
+  auto next_data = orig_data;
+  while (!node_list.empty()) {
+    auto cur_pair = node_list.back();
+    // Add input to current node.
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:375]: Focus on Loop in this PR, adding the verification first.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 2]
```diff
@@ -336,7 +336,64 @@ void SquashSliceAndSelect(Node* index_put_node) {
   new_index_put->copyMetadata(index_put_node->output());
   index_put_node->output()->replaceAllUsesWith(new_index_put);
 
-  orig_data->replaceAllUsesAfterNodeWith(new_index_put->node(), new_index_put);
+  auto block_node = new_index_put->node();
+  auto outer_block = block_node->owningBlock();
+  auto next_node = outer_block->owningNode();
+  if (nullptr == next_node) {
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:342]: Done.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 3]
```diff
@@ -336,7 +336,64 @@ void SquashSliceAndSelect(Node* index_put_node) {
...
+    orig_data->replaceAllUsesAfterNodeWith(
+        new_index_put->node(), new_index_put);
+    return;
+  }
+  for (auto block_inputs : outer_block->inputs()) {
+    if (block_inputs->debugName() == orig_data->debugName()) {
+      AT_ERROR(
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:349]: We add `orig_data` to `block_input` when the first index_put logic comes. If there is a second one, then the orig_data in the second one will match index_put.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 4]
```diff
@@ -336,7 +336,64 @@ void SquashSliceAndSelect(Node* index_put_node) {
...
+    }
+  }
+
+  // Register index_put outputs through the blocks.
+  outer_block->registerOutput(new_index_put);
+  std::vector<std::pair<Block*, Node*>> node_list = {
+      std::make_pair(outer_block, next_node)};
+  next_node->addOutput()->copyMetadata(new_index_put);
+  auto next_block = next_node->owningBlock();
+  while (true) {
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:360]: Good catch.

File: torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp [hunk 5]
```diff
@@ -336,7 +336,64 @@ void SquashSliceAndSelect(Node* index_put_node) {
...
+  }
+  for (auto block_inputs : outer_block->inputs()) {
+    if (block_inputs->debugName() == orig_data->debugName()) {
+      AT_ERROR(
+          "More than one aten::index_put in a subblock are not supported.");
+    }
+  }
+
+  // Register index_put outputs through the blocks.
+  outer_block->registerOutput(new_index_put);
```
[the remark this answers is not in the document]
Comment (jiafatom) [reply torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp:355]: Add a check: if new_index_put is already in output, just return.

**Comment by jiafatom** [comment]
> some inline comments. It would be helpful if each step there is some comments with simple IR snippets showing before and after comparison

IR snippets added.

**Review by BowenBao — APPROVED** [review]
LGTM. CI failures look unrelated.

**Comment by jiafatom** [comment]
@bzinodev Could you please sign off this PR? Two huggingface model conversion depends on this PR. Thank you!

**Review by bzinodev — APPROVED** [review]

**Comment by jiafatom** [comment]
@bzinodev May I know when this PR can get into master? It is imported to Phabricator for 5 days. THank you!

**Comment by jiafatom** [comment]
@SplitInfinity  This PR has been in Phabricator for 11 days. May I know when can we merge this PR to master? Who shall I contact with? Thank you!
cc @bzinodev I understand you are probably busy with other things. Thanks!

**Comment by jiafatom** [comment]
@bzinodev I just had a meeting with my team. They mentioned that you are the contact point to merge this PR since this PR is approved before the new policy is in effect. Could you please merge it to master? 
cc @spandantiwari

**Pull request closed without merging**